src/platforms/hosted/bmda_gpiod.c

```diff
@@ -0,0 +1,264 @@
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/* Implement libgpiod based GPIO backend */
+
+#define _POSIX_C_SOURCE 200809L
+
+#include <gpiod.h>
+#include "general.h"
+#include <errno.h>
+#include <string.h>
+#include <limits.h>
+#include <stdbool.h>
+
+#include "bmda_gpiod.h"
+
+struct gpiod_line *bmda_gpiod_tck_pin;
+struct gpiod_line *bmda_gpiod_tms_pin;
+struct gpiod_line *bmda_gpiod_tdi_pin;
+struct gpiod_line *bmda_gpiod_tdo_pin;
+
+bool bmda_gpiod_jtag_ok = false;
+
+struct gpiod_line *bmda_gpiod_swdio_pin;
+struct gpiod_line *bmda_gpiod_swclk_pin;
+
+bool bmda_gpiod_swd_ok = false;
+
+uint32_t target_clk_divider = UINT32_MAX;
+
+static void bmda_gpiod_debug_pin(struct gpiod_line *line, const char *op, bool print, bool val)
+{
+#ifdef DEBUG
+	DEBUG_WIRE("GPIO %s %s", gpiod_line_consumer(line), op);
+	if (print)
+		DEBUG_WIRE("=%d", val);
+	DEBUG_WIRE("\n");
+#else
+	(void)line;
+	(void)op;
+	(void)print;
+	(void)val;
+#endif
+}
+
+void bmda_gpiod_set_pin(struct gpiod_line *pin, bool val)
+{
+	if (pin) {
+		bmda_gpiod_debug_pin(pin, "set", true, val);
+		if (gpiod_line_set_value(pin, val ? 1 : 0)) {
+			DEBUG_ERROR("Failed to set pin to value %d errno: %d", val, errno);
+			exit(1);
+		}
+	} else {
+		DEBUG_ERROR("BUG! attempt to write uninit GPIO");
+	}
+}
+
+bool bmda_gpiod_get_pin(struct gpiod_line *pin)
+{
+	if (pin) {
+		int ret = gpiod_line_get_value(pin);
+		if (ret < 0) {
+			DEBUG_ERROR("Failed to get pin value errno: %d", errno);
+			exit(1);
+		}
+		bmda_gpiod_debug_pin(pin, "read", true, !!ret);
+		return !!ret;
+	} else {
+		DEBUG_ERROR("BUG! attempt to read uninit GPIO");
+		exit(1);
+	}
+}
+
+void bmda_gpiod_mode_input(struct gpiod_line *pin)
+{
+	if (pin) {
+		bmda_gpiod_debug_pin(pin, "input", false, false);
+		if (gpiod_line_set_direction_input(pin)) {
+			DEBUG_ERROR("Failed to set pin to input errno: %d", errno);
+			exit(1);
+		}
+	} else {
+		DEBUG_ERROR("BUG! attempt to set uninit GPIO to input");
+	}
+}
+
+void bmda_gpiod_mode_output(struct gpiod_line *pin)
+{
+	if (pin) {
+		bmda_gpiod_debug_pin(pin, "output", false, false);
+		if (gpiod_line_set_direction_output(pin, 0)) {
+			DEBUG_ERROR("Failed to set pin to output errno: %d", errno);
+			exit(1);
+		}
+	} else {
+		DEBUG_ERROR("BUG! attempt to set uninit GPIO to output");
+	}
+}
+
+static bool bmda_gpiod_parse_gpio(const char *name, char *gpio)
+{
+	DEBUG_INFO("GPIO set %s: %s\n", name, gpio);
+	char *offset = strstr(gpio, ":");
+	if (!offset)
+		return false;
+	*offset = '\0';
+	offset++;
+	char *end = offset + strlen(offset);
+	char *valid;
```

Comment:
Please initialise this to `NULL` as strtoul() may fail and not touch this on some libc's, resulting in UB read-before-init in the if check below.

Reply:
I think I have sorted this out.